### PR TITLE
Made testing at max range resilient to floating point rounding errors.

### PR DESCRIPTION
**Pull request by martinbudden**
Also fixed error in DECIDEGREES_TO_RADIANS macro. See also:

https://github.com/cleanflight/cleanflight/issues/1599

### CODE REVIEW

File: src/test/unit/sonar_unittest.cc [hunk 1]
```diff
@@ -102,9 +102,11 @@ TEST(SonarUnittest, TestAltitude)
     // distance 400, 22.4 degrees of roll, this corresponds to HC-SR04 effective max detection angle
     EXPECT_EQ(sonarCalculateAltitude(400, cosf(DECIDEGREES_TO_RADIANS(224))), 369);
     EXPECT_EQ(sonarGetLatestAltitude(), 369);
-    // max range, max tilt
-    EXPECT_EQ(sonarCalculateAltitude(sonarMaxRangeCm, cosf(DECIDEGREES_TO_RADIANS(sonarMaxTiltDeciDegrees))), sonarMaxAltWithTiltCm);
-    EXPECT_EQ(sonarGetLatestAltitude(), sonarMaxAltWithTiltCm);
+    // Check limits at max range max tilt, need to deal with rounding errors in cosf calculation
+    EXPECT_GE(sonarCalculateAltitude(sonarMaxRangeCm, cosf(DECIDEGREES_TO_RADIANS(sonarMaxTiltDeciDegrees - 1))), sonarMaxAltWithTiltCm - 1);
```
Comment (ledvinap) [inline src/test/unit/sonar_unittest.cc]: there is `EXPECT_NEAR(val1, val2, abs_error);` macro..

https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md

**Comment by martinbudden** [comment]
@ledvinap Thanks. Now fixed.

**Comment by AndersHoglund** [comment]
Wowowow, hold it. This was already covered in PR1601 ( and Issue 1599). 
You should NOT do both. This PR adapts the test to a edge behaviour in the product code, whereas the PR1601 fixes the edge case. Doing both will make the test meaningless, as the test then does not test at the edge anymore, it is "-1" step off the edge.. There are already a test for the "-1" situation. Might as well remove the test.....
/A

**Comment by martinbudden** [comment]
I'm afraid I disagree. It's more subtle than that.

PR1601 fixes a bug in the code, namely an equality comparision of a floating point number. What is being compared is cosines, so the value will never be a binary fraction, so equality can never occur, so the equality comparison is simply wrong. As you point out using `<=` on floats is less dangerous, but it is still incorrect.

In the test code the tests:
` EXPECT_EQ(sonarCalculateAltitude(400, cosf(DECIDEGREES_TO_RADIANS(224))), 369);`
and
` EXPECT_NEAR(sonarCalculateAltitude(sonarMaxRangeCm, cosf(DECIDEGREES_TO_RADIANS(sonarMaxTiltDeciDegrees - 1))), sonarMaxAltWithTiltCm, 1);`
are not the same

The first tests the specific case of 22.4 degrees Tilt.
The second tests that `sonarMaxTiltDeciDegrees` and `sonarMaxAltWithTiltCm` values correspond to each other. This is important because `sonarMaxTiltDeciDegrees` is a property of th HC-SR04 device and its value is outside that in the spec sheet (the spec sheet gives 15 degrees). So this is a value that could be changed if a user goes out and performs some tests (indeed, I have thought of doing this myself at some point). So suppose someone does some measurements, obtains a better value for max tilt and then puts in a pull request with a changed value of max tilt. This test checks that the sonar initialisation code correctly calculates the value of `sonarMaxAltWithTiltCm`. I agree that it is rather unlikely that this will be done incorrectly, but that is not a reason not to test it. One of the purposes of test code is to prevent future mistakes, and this guards against an erroneous change in the sonar initialisation code.

And finally, we cannot use:
` EXPECT_NEAR(sonarCalculateAltitude(sonarMaxRangeCm, cosf(DECIDEGREES_TO_RADIANS(sonarMaxTiltDeciDegrees))), sonarMaxAltWithTiltCm, 1);`
since there is a discontinuity in `sonarCalculateAltitude`, once max tilt is exceeded `sonarCalculateAltitude` returns `SONAR_OUT_OF_RANGE` (-1), so the near test does not work right on the edge, so we have to step back from the edge and perform the test.

So this is not a mistake. It has been thoroughly thought through.

**Comment by AndersHoglund** [comment]
OK, I'm with you. For the most part. In fact the old ( or rather current ) test code did catch the equality bug (if only in some env.). The new test code will not. I just applied that patch to the test code, while keeping the original product code. And it passes. Does not catch the "<=" edge case bug. 
No big deal, in practise just a fractional difference, but it is an issue of testing principle.

If you only remove the little annoying "-1" thing, it will catch the bug. And the test will pass when bug is fixed. Yes, I have tried it.
I.e. there is a need for a test on the edge, not "-1 away from the edge." I can buy all the other changes in the test code.

**Comment by martinbudden** [comment]
I agree that there is a need for a test on the edge. However I am not convinced that a test right on the edge will pass reliably, that's why I've implemented the "-1 away from the edge". Yes it is desirable to have the test right on the edge. But it is also desirable that people in future don't spend time tracking down why the sonar test code has started failing.

**Comment by AndersHoglund** [comment]
OK, great. Only a test on the edge would have found the "<=" comparison. 
And we need that for regression testing. So no people in the future makes the same mistake.
And it is reliable. More reliable than the current. So, what are we arguing about ;-) 
/A

**Comment by AndersHoglund** [comment]
One suggestion and compromise if you still have doubts on stability. 
Lets do it in two steps:
  1. Merge #1601 only. Let things run for a while with the current tests. Put this PR on hold.
  2. If any problems shows up with this test, merge this PR. And I will be in shame...
OK?

**Comment by martinbudden** [comment]
Well, I agree with your comment that we are essentially arguing over nothing.

OK, let's go with the two step approach. Does that mean I should close this PR?

**Comment by AndersHoglund** [comment]
Might be the best way to close it. I have not found any way to put a PR on hold as I thought. But keep your changes and your task branch. Then it's easy to re-submit a new PR if needed.

**Comment by hydra** [comment]
I merged #1601
